When would an error "Cannot find context with specified id undefined" happen?

I am getting the following error:

```js
Error: Protocol error (Runtime.callFunctionOn): Cannot find context with specified id undefined
    at Session._onMessage (/srv/node_modules/puppeteer-edge/lib/Connection.js:205:25)
    at Connection._onMessage (/srv/node_modules/puppeteer-edge/lib/Connection.js:105:19)
    at emitOne (events.js:115:13)
    at WebSocket.emit (events.js:210:7)
    at Receiver._receiver.onmessage (/srv/node_modules/ws/lib/WebSocket.js:143:47)
    at Receiver.dataMessage (/srv/node_modules/ws/lib/Receiver.js:389:14)
    at Receiver.getData (/srv/node_modules/ws/lib/Receiver.js:330:12)
    at Receiver.startLoop (/srv/node_modules/ws/lib/Receiver.js:165:16)
    at Receiver.add (/srv/node_modules/ws/lib/Receiver.js:139:10)
    at Socket._ultron.on (/srv/node_modules/ws/lib/WebSocket.js:139:22)
    at emitOne (events.js:115:13)
    at Socket.emit (events.js:210:7)
    at addChunk (_stream_readable.js:266:12)
    at readableAddChunk (_stream_readable.js:253:11)
    at Socket.Readable.push (_stream_readable.js:211:10)
    at TCP.onread (net.js:585:20)

```

It is a large codebase and it is unclear whats triggering this error.

Any guides?

On that note, there needs to be a better way to throw error. Without knowing the origin of the error in the code it is impossible to trace down these exceptions.

Hey folks, are we sure that `runtime.clearExecutionContexts` is doing the right thing? 

In the code it clears out all execution contexts in all frames. The reason I ask is that I'm also running into this in a different library. It doesn't happen all the time, only occassionally.

Here's a full feed of the actions (most of these logs correspond to puppeteer actions):

```
=== RUN   TestClick
DevTools listening on ws://127.0.0.1:53553/devtools/browser/224356af-06f2-4a9d-ba05-af67173d5d12
  INFO[0023] set main frame                 id=9612C9BCADCD1D4AAA6C9241958BDAD4 parent=<nil>
  INFO[0023] chrome/page/goto: navigating url=https://google.com
  INFO[0024] chrome/frame/waituntil: still waiting for events events=unload id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0024] chrome/frame/setlifecycle: setting lifecycle event=init id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0024] chrome/frame/setlifecycle: set navigating channel
  INFO[0024] chrome/frame/clearexecutioncontext: cleared all execution contexts frame=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0024] chome/page: execution context destroyed: couldn't find the frame id execution=1
  INFO[0024] chrome/frame/addexecutioncontext: added an execution context frame=9612C9BCADCD1D4AAA6C9241958BDAD4 id=2 is_default=true
  INFO[0024] chrome/frame/setlifecycle: setting lifecycle event=firstPaint id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0024] chrome/frame/setlifecycle: setting lifecycle event=firstContentfulPaint id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0024] chrome/frame/setlifecycle: setting lifecycle event=firstTextPaint id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0024] chrome/frame/setlifecycle: setting lifecycle event=firstImagePaint id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0024] chrome/frame/setlifecycle: setting lifecycle event=firstMeaningfulPaintCandidate id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0024] chrome/frame/setlifecycle: setting lifecycle event=firstMeaningfulPaintCandidate id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0024] chrome/frame/setlifecycle: setting lifecycle event=DOMContentLoaded id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0024] chrome/frame/waituntil: event found event=unload id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0024] chrome/frame/waituntil: event found event=DOMContentLoaded id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0024] chrome/frame/waituntil: still waiting for events events=load id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0024] chrome/frame/setlifecycle: setting lifecycle event=load id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0025] chrome/frame/waituntil: event found event=load id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0025] chrome/frame/waituntil: still waiting for events events=unload, DOMContentLoaded, load id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0025] chrome/frame/setlifecycle: setting lifecycle event=init id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0025] chrome/frame/setlifecycle: set navigating channel
  INFO[0025] chrome/frame/setlifecycle: setting lifecycle event=firstMeaningfulPaint id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0025] chrome/frame/setlifecycle: setting lifecycle event=networkAlmostIdle id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0025] chrome/frame/setlifecycle: setting lifecycle event=networkIdle id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0025] chrome/frame/waituntil: event found event=unload id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0025] chrome/frame/waituntil: still waiting for events events=DOMContentLoaded, load id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0025] chrome/frame/setlifecycle: setting lifecycle event=init id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0025] chrome/frame/setlifecycle: set navigating channel
  INFO[0026] chrome/frame/removeexecutioncontext: removing execution context id=2
  INFO[0026] chrome/frame/addexecutioncontext: added an execution context frame=9612C9BCADCD1D4AAA6C9241958BDAD4 id=3 is_default=true
  INFO[0026] chrome/frame/clearexecutioncontext: cleared all execution contexts frame=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0026] chrome/frame/setlifecycle: setting lifecycle event=firstPaint id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0026] chrome/frame/setlifecycle: setting lifecycle event=firstMeaningfulPaintCandidate id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0026] chrome/frame/setlifecycle: setting lifecycle event=firstContentfulPaint id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0026] chrome/frame/setlifecycle: setting lifecycle event=firstTextPaint id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0026] chrome/frame/setlifecycle: setting lifecycle event=DOMContentLoaded id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0026] chrome/frame/setlifecycle: setting lifecycle event=firstImagePaint id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0026] chrome/frame/setlifecycle: setting lifecycle event=load id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0026] chrome/frame/setlifecycle: setting lifecycle event=firstMeaningfulPaintCandidate id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0026] chrome/frame/setlifecycle: setting lifecycle event=firstMeaningfulPaintCandidate id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0027] chrome/frame/waituntil: event found event=DOMContentLoaded id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0027] chrome/frame/waituntil: event found event=load id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0027] chrome/frame/setlifecycle: setting lifecycle event=firstMeaningfulPaint id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0027] chrome/frame/setlifecycle: setting lifecycle event=networkAlmostIdle id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0027] chrome/frame/setlifecycle: setting lifecycle event=networkIdle id=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0027] chrome/page/title: getting title of main frame
  INFO[0027] chrome/frame/defaultexecutioncontext: getting the default execution context id=9612C9BCADCD1D4AAA6C9241958BDAD4

--- FAIL: TestClick (8.56s)
	chrome_test.go:106: chrome/page: unable to get the default execution context: headless/retrier: timed out: chrome/frame: couldn't find the default execution context
```

From the looks of it, it seems like the most important bit is here:

```
  INFO[0026] chrome/frame/removeexecutioncontext: removing execution context id=2
  INFO[0026] chrome/frame/addexecutioncontext: added an execution context frame=9612C9BCADCD1D4AAA6C9241958BDAD4 id=3 is_default=true
  INFO[0026] chrome/frame/clearexecutioncontext: cleared all execution contexts frame=9612C9BCADCD1D4AAA6C9241958BDAD4
```

**Oh!** You know I bet it's a race condition between adding new execution contexts and clearing out previous execution contexts 😀

The order should probably be:

```
  INFO[0026] chrome/frame/removeexecutioncontext: removing execution context id=2
  INFO[0026] chrome/frame/clearexecutioncontext: cleared all execution contexts frame=9612C9BCADCD1D4AAA6C9241958BDAD4
  INFO[0026] chrome/frame/addexecutioncontext: added an execution context frame=9612C9BCADCD1D4AAA6C9241958BDAD4 id=3 is_default=true
```

**Follow up**, do we even need to clear the executions? Or can we just let `runtime.executionContextDestroyed` do it's thing?

--- a/lib/FrameManager.js
+++ b/lib/FrameManager.js
@@ -869,7 +869,9 @@ class WaitTask {
     }
 
     // Ignore timeouts in pageScript - we track timeouts ourselves.
-    if (!error && await this._frame.evaluate(s => !s, success)) {
+    // If the frame's execution context has already changed, `frame.evaluate` will
+    // throw an error - ignore this predicate run altogether.
+    if (!error && await this._frame.evaluate(s => !s, success).catch(e => true)) {
       await success.dispose();
       return;
     }

--- a/test/frame.spec.js
+++ b/test/frame.spec.js
@@ -84,6 +84,14 @@ module.exports.addTests = function({testRunner, expect}) {
       await page.evaluate(() => window.__FOO = 1);
       await watchdog;
     });
+    it('should work when resolved right before execution context disposal', async({page, server}) => {
+      await page.evaluateOnNewDocument(() => window.__RELOADED = true);
+      await page.waitForFunction(() => {
+        if (!window.__RELOADED)
+          window.location.reload();
+        return true;
+      });
+    });
     it('should poll on interval', async({page, server}) => {
       let success = false;
       const startTime = Date.now();

